src/storage-operators/src/s3_oneshot_sink.rs

```diff
@@ -56,33 +61,301 @@ pub fn copy_to<G, F>(
+            .await?;
+            Ok::<(), anyhow::Error>(())
+        };
+
+        if is_leader {
+            let res = leader_work().await.map_err(|e| e.to_string());
+            start_handle.give(&start_cap, res).await;
+        };
+    });
+
+    start_stream
+}
+
+/// Renders the 'completion' operator, which expects a `completion_stream`
+/// that contains the result from the upload operations to return to the
+/// `onetime_callback` (see description of `copy_to`).
+///
+/// On the leader worker, this operator waits to see the empty frontier for
+/// the completion_stream and then does some cleanup work before returning
+/// the callback.
+///
+/// This cleanup work removes the INCOMPLETE sentinel file (see description
+/// of `render_initialization_operator` for more details).
+fn render_completion_operator<G, F>(
+    scope: G,
+    connection_context: ConnectionContext,
+    aws_connection: AwsConnection,
+    connection_id: GlobalId,
+    sink_id: GlobalId,
+    s3_key_manager: S3KeyManager,
+    completion_stream: Stream<G, Result<u64, String>>,
+    onetime_callback: F,
+) where
+    G: Scope<Timestamp = Timestamp>,
+    F: FnOnce(Result<u64, String>) -> () + 'static,
+{
     let worker_id = scope.index();
-    let is_leader = usize::cast_from((sink_id, "leader").hashed()) % scope.peers() == worker_id;
+    let num_workers = scope.peers();
+    let leader_id = usize::cast_from((sink_id, "completion").hashed()) % num_workers;
+    let is_leader = worker_id == leader_id;
+
+    let mut builder = AsyncOperatorBuilder::new("CopyToS3-completion".to_string(), scope.clone());
+
+    let mut completion_input = builder.new_disconnected_input(&completion_stream, Pipeline);
 
-    let mut builder = AsyncOperatorBuilder::new("CopyToS3".to_string(), scope.clone());
+    builder.build(move |_| async move {
+        // fallible async block to use the `?` operator for convenience
+        let leader_cleanup_work = || async move {
+            info!(%sink_id, %worker_id, "s3 leader worker completion");
+            let sdk_config = aws_connection
+                .load_sdk_config(&connection_context, connection_id, InTask::Yes)
+                .await?;
+
+            let client = mz_aws_util::s3::new_client(&sdk_config);
+
+            // Remove the INCOMPLETE sentinel file to indicate that the upload is complete.
+            // This will race against other replicas who are completing the same uploads,
+            // such that the first replica to complete its uploads will delete the sentinel
+            // and the subsequent replicas shouldn't error if the object is already deleted.
+            // TODO: Should we also write a manifest of all the files uploaded?
+            mz_ore::task::spawn(|| "copytos3:completion", {
+                let client = client.clone();
+                let bucket = s3_key_manager.bucket.clone();
+                let incomplete_sentinel_key = s3_key_manager.incomplete_sentinel_key();
```

Comment:
Same tip here of hoisting let bindings and using `async move` without a surrounding block